### ISSUE TITLE
How to use drive_auth non-interactively?

### ISSUE THREAD

**Issue by abeburnett** [tidyverse/googledrive#239]
Whenever I run the following:

```r
googledrive::drive_auth(path = "/home/folders/googlesheets_token.json", cache = TRUE)
```

I get the following console output (with a 1: SOME_OPTION): 

```text
The googledrive package is requesting access to your Google account.
Select a pre-authorised account or enter '0' to obtain a new token.
Press Esc/Ctrl + C to abort.
```

The issue is that this script is designed to run on a scheduler; i.e., not interactively. How in the heck can I do the token load and/or refresh when necessary when I'm not running these scripts interactively?

Please advise!

**Comment by jennybc** [comment]
How exactly did you create `/home/folders/googlesheets_token.json`? The use of `path` indicates you are using a service account token. But the `cache` argument only applies to OAuth tokens. So things are rather jumbled. The name contains "googlesheets" which also makes me wonder which APIs are enabled for this service token (if it is a service token).

What version of googledrive do you have? CRAN or a development version from here on GitHub?

**Comment by abeburnett** [comment]
Hi Jenny! It could very well be a bit jumbled since I'm switching over from your excellent `googlesheets` package to your also excellent `googledrive` package. I'm not entirely clear on what the good practices are for authentication flows with the googledrive package. Using googlesheets I needed to save the token explicitly, so my assumption was that I needed to do that with the googledrive package as well. Is this assumption incorrect?

I'm using the development version of googledrive from GitHub: googledrive_0.1.2.9000

Thanks for your rapid response!

**Comment by jennybc** [comment]
The dev version of googledrive has switched to a new auth system and it is undoubtedly not well documented yet.

I don't know of any really compelling features that only exists in the GitHub version.

Unless you have a reason to do otherwise, I suggest you use the CRAN version of googledrive.

And then this documentation should hold and, yes, it works pretty similar to googlesheets.

https://googledrive.tidyverse.org/reference/drive_auth.html

Let me know if you have trouble.

The auth will be slightly different in the next release of googledrive, as we work out a common system across multiple packages (for Drive, Sheets, Gmail, etc.). But that's not released yet.

**Comment by TetricAttack** [comment]
Hi Jenny, I hope you are doing well. 

I'm also unable to get googledrive to work un-interactively by running it on R Markdown. I created a token using this procedure stated on the reference below.

`## store token in an object and then to file
ttt <- drive_auth()
saveRDS(ttt, "token.rds")`

Then, when I call  `drive_auth(oauth_token = "token.rds")` on the R markdown and use `drive_get` right after I get the following interactive message that asks to select the desired account by pressing 1 (which is the part that I want to skip)

`The googlesheets4 package is requesting access to your Google account.Select a pre-authorised account or enter '0' to obtain a new token.Press Esc/Ctrl + C to abort.`

Any help is appreaciated! Thank you.

**Comment by jennybc** [comment]
@TetricAttack Can you put a chunk in the Rmd to reveal:

  * version of googledrive (`packageVersion("googledrive")`)
  * listing of current working directory (`list.files()`)
  * output of `file.exists("token.rds")`

And then use `knitr::knit_exit()` so that the rest of the R code isn't executed?
Then lets look at the result to try to figure out why this token isn't being put into force.

**Comment by jennybc** [comment]
Hold on:

> googlesheets4 package

This repo is googledrive. But you seem to be calling googlesheets4.

**Comment by TetricAttack** [comment]
You are correct, I was mixing the two packages and my Rmd is actually failing to authenticate with `googlesheets4`. 

Somewhere in my mind I thought the auth process on `googlesheets4` depended on `googledrive`. Just looked up and the `googlesheets4` auth is still work in progress. I'll keep an eye on the update! Thank you for your answers!

**Comment by jennybc** [comment]
Yeah unfortunately their auth processes are separate/different at the moment. But they should both work.

**Comment by DataStrategist** [comment]
Adding my experience with this. It may be helpful to someone. I saw that everything was broken, so these are the steps I took:

 1. Since I"m running this on a server (which won't authenticate due to a redirect) , I clone the repo locally.
 2. Authenticate on my computer, SAVE the file, and then copy that to the correct location on the server (in my case it was: `/home/amit/.R/gargle/gargle-oauth/d9443dbd06_changed_of_course_34joihf2f92_XXXXX@gmail.com`)
 3. I was still having trouble running this without the prompt that still asked me to confirm if that was the auth I wanted to use (perhaps if there's only one it shouldn't ask?)
 4. As Jenny said, I consulted [the new doc](https://googledrive.tidyverse.org/reference/drive_auth.html) and found `drive_auth(oauth_token = blah)`. Which didn't work because the input `oauth_token` doesn't exist in the dev version (as of today). But in the documentation it clearly says that this applies to the CRAN version... so I just reinstalled the CRAN version and all works well now.

Good job Jenny!

**Comment by jennybc** [comment]
Thanks @DataStrategist for the detailed report.

I would recommend using CRAN googledrive still, for smoothest operations.

But if one is willing to put up with dev hiccups and provide feedback, then using dev googledrive + (therefore) dev gargle, that does help me with development & testing.

---
I believe these points describe behaviour in dev googledrive + dev gargle.

> I was still having trouble running this without the prompt that still asked me to confirm if that was the auth I wanted to use

To avoid this, explicitly do `drive_auth(email = "XYZ@example.com")`, which we regard as giving permission to use an existing token.

> perhaps if there's only one it shouldn't ask?

We're trying to be very careful about NOT automatically reading/using/creating tokens and, at this point, we've decided this still is too implicit.

---
> I consulted the new doc and found `drive_auth(oauth_token = blah)`.

Now this describes CRAN googledrive.

---
> EDIT - I spoke too soon

Then it feels like we're back to dev googlerive + dev gargle.

--

I think it would be helpful, for this discussion and your success, to be super explicit about the version of gargle and googledrive installed both locally and on the server. And to make sure they are the same.

So either CRAN googledrive. Period.
Or dev googledrive + dev gargle.

**Comment by DataStrategist** [comment]
Thanks for all of that Jenny. 

the `email` input isn't available in my version of googledrive (the CRAN version). Should it be?

To be clear about versions, here's my session:

```
R version 3.4.4 (2018-03-15)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 16.04.6 LTS

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.6.0
LAPACK: /usr/lib/lapack/liblapack.so.3.6.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] readr_1.1.1       googledrive_0.1.3

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.1       crayon_1.3.4     dplyr_0.8.0.1    assertthat_0.2.1 R6_2.4.0         magrittr_1.5    
 [7] pillar_1.3.1     rlang_0.3.4      rstudioapi_0.8   tools_3.4.4      glue_1.3.1       purrr_0.3.2     
[13] hms_0.4.2        yaml_2.2.0       compiler_3.4.4   pkgconfig_2.0.2  tidyselect_0.2.5 tibble_2.1.1    
```

I don't see gargle above, it's version 0.0.0.9004.

Happy to run any additional diagnostics.

**Comment by jennybc** [comment]
> the email input isn't available in my version of googledrive (the CRAN version). Should it be?

Nope that is exactly as expected. The `email` argument is only in the dev version and is related to delegating auth to gargle.

Sorry this is so confusing at this stage. I'd advise either sticking to CRAN everything or ... steeling yourself for this situation where things are a bit changeable and have poor documentation.

**Comment by DataStrategist** [comment]
You're apologizing to me? Don't be silly! I'm the one that feels weird that I can't give you clear feedback. 

Anyway, I am using only the CRAN version... so as of right now, I guess the best way to deal w/ the error is the way I have done it above, calling `WorksheetsThrowaway <-   safely(drive_find)()` the first time? I would switch to the dev version to help test that out but I am presenting this bit of functionality today (I'm visiting my HQ) so can't mess around too much! Next week, I can test everything out on the dev branch!

WE GOT THIS! 💪 ⭐️ 🌟

**Comment by jennybc** [comment]
I find the need to call `drive_find()` very strange.

Can you revisit a detailed account of what happens, now that we've cleared up the CRAN vs dev googledrive thing? So a version of the story that makes no mention of gargle and does not attempt to use the `email` argument.

**Comment by DataStrategist** [comment]
Sure. Sorry for the delay.

So when I call googledrive (the CRAN version) I first have to authenticate, so I do that with `drive_auth(oauth_token = blah)`. This is the token that I had to build in my laptop and then copy over to the linux server because I can't authenticate on the server due to the redirect not being able to call the Linux Rstudio in my browser. 

Then, whenever I call a googledrive function (any googledrive function), the first time, it errors out quoting a stale token:

```
Auto-refreshing stale OAuth token.
Error in gzfile(file, mode) : cannot open the connection
In addition: Warning message:
In gzfile(file, mode) :
  cannot open compressed file 'C:/Users/Amit/.R/gargle/gargle-oauth/d9443dbd067-EDITED-60d52_XXXXXX@gmail.com', probable reason 'No such file or directory'
```

The fact that it's quoting `C:/Users/Amit` maybe implies that somewhere that token remembers that it was done in my laptop so it's freaking out about that? 

However it only errors out the first time. So the first time I run the function, I wrap it in `safely()` and throw that away, then repeat the operation and all works as planned. 

Writing this I realize that perhaps the original problem and best potential solution is to give me the ability to authenticate FROM THE SERVER _per se_. Not sure that's possible.

Is this what you wanted? Standing by.

**Comment by jennybc** [comment]
This

```
C:/Users/Amit/.R/gargle/gargle-oauth/ ...
```

suggests we are still in some muddy place with CRAN vs dev version of googledrive. In a setting where only CRAN googledrive had ever been used, there will be no mention of gargle, no attempt to store a token in this directory up at the user-level, etc. So you're still somehow stuck between the two worlds.

**Comment by DataStrategist** [comment]
Interesting... ok thanks for that. Let me try uninstalling all 3: googledrive, googlesheets, and gargle. I'll give it another whirl in the morning. Thanks for your infinite help!

**Comment by DataStrategist** [comment]
ffs! My account is "acting suspicious" so they want text verification of my account but I'm travelling and can't get text messages!

Sorry Jenn, I'll feed back more next week when I'm back in London. Sorry

**Comment by DataStrategist** [comment]
@jennybc , just to confirm, using CRAN version of everything solves all problems! Thanks and sorry about the confusion. 👍

**Comment by EdenAnteby** [comment]
Sorry for the interruption, regarding - 

>" makes me wonder which APIs are enabled for this service token (if it is a service token)"

 - Which API should be enabled for a google service account in order to use googledrive package?
Is it just Drive API?

 - Also, I know it's not 100% the right place, but if it's ok, I will throw in another question - 
when you write 
>path = "*...token.json"

This file is the credentials file from google API console? if not, how can I get the correct token?

Thanks!

**Comment by jennybc** [comment]
> Which API should be enabled for a google service account in order to use googledrive package?
Is it just Drive API?

Yes.

> This file is the credentials file from google API console? if not, how can I get the correct token?

There are (at least) 2 types of credential-y files that you could download from <https://console.cloud.google.com/project> as JSON:

  * OAuth2 credentials, in the sense of a client and secret, that you then use in the process of doing 3-legged auth, to get OAuth2 token. In the httr/R world, we call this an "oauth app".  googledrive has a built in app but users can bring their own using `drive_auth_config()` (current version) or `drive_auth_configure()` (preferred name, already in the dev version).
  * Service account token. This is an actual token and this is what we're talking about in the `drive_auth(path = "/path/to/your/service/account/token.json"). Once loaded, it is used to obtain the same type of perishable token as we get with user OAuth tokens.

**Comment by jennybc** [comment]
BTW, I just cleared a huge hurdle getting our main GCP project verified yesterday, so our OAuth consent screen has a logo and a link to our Privacy Policy and we are not at risk of getting suspended due to too many unique users for a unverified app. This unblocks me from re-releasing gargle and a new version of googledrive. So the path is finally clear for a googledrive update.

**Comment by EdenAnteby** [comment]
That's great. I wanted to make sure I understand the steps before integrating it in my shiny app.
Many thanks!

**Comment by jennybc** [comment]
I think Shiny still represents a more challenging place to do all of this. Joe Cheng conducted a recent experiment, which I also worked on, re: a possible design pattern. And it worked. But it also lead to the conclusion that it would be nice to make a few small changes in Shiny, I think, to really make things slick. So stay tuned.

I say this just so you know that I know that Shiny has not gotten any special attention yet.

**Comment by EdenAnteby** [comment]
Oh, I probably missing something. 
If In my shiny server I'll point - drive_auth() to use a Google service account token (let's say the JSON file will be in my WD), what else do I need in order to be able to use the googledrive package? 
My final goal is to read a google sheet file automatically.

**Comment by jennybc** [comment]
If this is a sheet you are willing to make readable by "Anyone with a link", I would do that and not use auth at all. Use `drive_deauth()` and wash your hands of the problem of managing a token.

But if you can't do that, then what you say will work. If the app just needs to read the sheet, I would try to scope the token for read only access, since you're going to have to put it on a server.

Things only get really fiddly if you're trying to facilitate auth by the *users* of your app, so that's what I was worried about.

**Comment by MostlyR** [comment]
Hi Jenny.
I am trying to create a scheduled job for data ingestion from a gsheet to a hive table.
To do so i am using airflow and i need to create a token file in R to access the file (not globally shared).

I am doing
ttt <- drive_auth(oauth_token = NULL, service_token = NULL, reset = TRUE,
                  cache = TRUE,
                  use_oob = TRUE, verbose = TRUE)
saveRDS(ttt, "ttt.rds")

to create the rds file and then calling it using
drive_auth ("full/path/to/folder/ttt.rds")
obtaining the following error:

Error in eval(parse(text = text, keep.source = FALSE), envir) :
INFO - object 'token' not found

Any advice?

**Comment by jennybc** [comment]
@MostlyR 

What version of googledrive are you using? I advise developing anything new with the development version, here on GitHub, because a release is imminent, and I'm modifying auth somewhat, to use the gargle package.

This vignette describes how to do non-interactive auth with any gargle-using package:

https://gargle.r-lib.org/articles/non-interactive-auth.html

That applies to the CRAN version of bigrquery, the dev version of googledrive, the dev version of googlesheets4. I'm gradually working through those CRAN releases and googledrive is next.

**Comment by MostlyR** [comment]
Hi @jennybc 
Thanks for your quick reply.

I am using v0.1.3 installed from the CRAN.
Unfortunately i am building a prototype for the company i work for, so i can't avoid building the automatisation piece.
Do you think googlesheets would work better in this context?

**Comment by jennybc** [comment]
> Unfortunately i am building a prototype for the company i work for, so i can't avoid building the automatisation piece.

I don't follow this. What does "automatisation piece" refer to?

> Do you think googlesheets would work better in this context?

Definitely avoid building anything new on googlesheets (the original package, the one that is on CRAN), as it is basically sunsetted. Strongly favour building around googledrive (preferably the dev version, soon to be updated) and googlesheets4 (only on GitHub, soon to have an official release, initially with only read capability).

**Comment by jennybc** [comment]
This has become a very long thread, with multiple story lines and describing a moving target (googledrive). I'm closing.

1. I think non-interactive auth is easier and better documented in the coming v1.0.0, which uses gargle. The gargle vignette [Non-interactive auth](https://gargle.r-lib.org/articles/non-interactive-auth.html) is the definitive guide now.
1. Any remaining problems with functionality or docs would be better discussed in a new issue thread

**Comment by guhanrv** [comment]
Hi @jennybc,

I am sorry if this is a repetitive question, but is there a way to generate a token RDS file like in the original `googlesheets` R package? Before, something like this used to work:

```
library(googlesheets)
token <- gs_auth(cache = FALSE)
gd_token()
saveRDS(token, file = "googlesheets_token.rds")
```
Is there an equivalent of this `saveRDS()` command for `googlesheets4`?
